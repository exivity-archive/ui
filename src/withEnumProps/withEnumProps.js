// @todo remove ThemeConsumer ponyfill when upgrading to styled-components v4

import React from 'react'
import PropTypes from 'prop-types'

class ThemeConsumer extends React.Component {
  render () {
    return this.context['__styled-components__next__']
      ? this.props.children(this.context['__styled-components__next__'].getTheme())
      : this.props.children({})
  }
}

ThemeConsumer.propTypes = {
  children: PropTypes.func.isRequired
}

ThemeConsumer.contextTypes = {
  '__styled-components__next__': PropTypes.shape({
    getTheme: PropTypes.func,
    subscribe: PropTypes.func,
    unsubscribe: PropTypes.func
  })
}

function withEnumProps (WrappedComponent, enumProps) {
  class WithEnumProps extends React.Component {
    render () {
      return <ThemeConsumer>
        {theme => {
          const compactedProps = {}
          Object.keys(enumProps).forEach(enumProp => {
            if (theme[enumProp]) {
              // Theme has this prop defined
              if (enumProps[enumProp] === 'key') {
                // We use the enumProp name (key)
                compactedProps[enumProp] = Object.keys(theme[enumProp])
                  .find(key => this.props[key])
              } else if (enumProps[enumProp] === 'value') {
                // We use the enumProp theme value
                const match = Object.entries(theme[enumProp])
                  .find(entry => this.props[entry[0]])
                compactedProps[enumProp] = match && match[1]
              } else {
                throw new Error('Invalid configuration')
              }
            }
          })

          return <WrappedComponent {...this.props} {...compactedProps} />
        }}
      </ThemeConsumer>
    }
  }

  WithEnumProps.propTypes = WrappedComponent.propTypes || {}
  WithEnumProps.defaultProps = WrappedComponent.defaultProps || {}
  WithEnumProps.displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component'

  return WithEnumProps
}

export default withEnumProps
