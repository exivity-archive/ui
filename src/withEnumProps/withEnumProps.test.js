/* eslint-disable no-undef */
import React from 'react'
import renderer from 'react-test-renderer'

import withEnumProps from './withEnumProps'

test('renders default WrappedComponent', () => {
  const WithEnumProps = withEnumProps(<div />)
  const component = renderer.create(<WithEnumProps />)
  expect(component.toJSON()).toMatchSnapshot()
})
