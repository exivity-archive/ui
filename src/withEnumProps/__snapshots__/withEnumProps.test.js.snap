// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default WrappedComponent 1`] = `null`;
